src/x86_64/unwind_rule.rs

```diff
@@ -19,6 +19,107 @@ pub enum UnwindRuleX86_64 {
     },
     /// (sp, bp) = (bp + 16, *bp)
     UseFramePointer,
+    /// sp = sp + 8 * (offset + register count)
+    /// This supports the common case of pushed callee-saved registers followed by a stack
+    /// allocation. Up to 8 registers can be stored, which covers all callee-saved registers (aside
+    /// from RSP which is implicit).
+    OffsetSpAndPopRegisters(UnwindRuleOffsetSpAndPopRegisters),
+}
+
+#[derive(Default, Clone, Copy, Debug, PartialEq, Eq)]
+pub struct UnwindRuleOffsetSpAndPopRegisters {
+    pub sp_offset_by_8: u16,
+    registers: u16,
+}
+
+pub enum OffsetOrPop {
+    None,
+    OffsetBy8(u16),
+    Pop(Reg),
+}
+
+impl UnwindRuleOffsetSpAndPopRegisters {
+    /// Get the rule which represents the given operations, if possible.
+    pub fn for_operations<I, T>(iter: I) -> Option<UnwindRuleX86_64>
+    where
+        I: Iterator<Item = T>,
+        T: Into<OffsetOrPop>,
+    {
+        let mut iter = iter.map(Into::into).peekable();
+        let mut rule = Self::default();
+        if let Some(OffsetOrPop::OffsetBy8(offset)) = iter.peek() {
+            rule.sp_offset_by_8 = *offset;
+            iter.next();
+        }
+        let mut regs = Vec::new();
+        for i in iter {
+            if let OffsetOrPop::Pop(reg) = i {
+                regs.push(reg);
+            } else {
+                return None;
+            }
+        }
+
+        if regs.is_empty() && rule.sp_offset_by_8 == 0 {
+            Some(UnwindRuleX86_64::JustReturn)
+        } else if rule.set_registers(&regs) {
+            Some(UnwindRuleX86_64::OffsetSpAndPopRegisters(rule))
+        } else {
+            None
+        }
+    }
+
+    /// Return the initial stack pointer offset, in bytes.
+    pub fn sp_offset(&self) -> u64 {
+        self.sp_offset_by_8 as u64 * 8
+    }
+
+    /// Return the ordered set of registers to pop.
+    pub fn registers(&self) -> Vec<Reg> {
+        use Reg::*;
+
+        let mut regs = vec![RBX, RBP, RDI, RSI, R12, R13, R14, R15];
```

Comment:
I haven't fully understood this yet; is this similar to the register permutation encoding that `OPCODE_KIND_X86_FRAMELESS_IMMEDIATE` uses? Could we reuse code from https://github.com/mstange/macho-unwind-info/blob/main/src/opcodes/permutation.rs ?

If this only supports one fixed order of registers, can you write a comment to that effect somewhere?

Reply:
Yes, it is the same algorithm to compress the registers. It supports one arbitrary order used for encode and decode, which I've chosen there (this is the ordering of callee-saved registers that the [microsoft docs](https://learn.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-170#callercallee-saved-registers) have).